src/cascadia/TerminalSettingsModel/Profile.cpp

```diff
@@ -533,5 +538,154 @@ Json::Value Profile::ToJson() const
     JsonUtils::SetValueForKey(json, BellStyleKey, _BellStyle);
     JsonUtils::SetValueForKey(json, PixelShaderPathKey, _PixelShaderPath);
 
+    JsonUtils::SetValueForKey(json, EnvironmentKey, _EnvironmentVariables);
+
     return json;
 }
+
+namespace
+{
+    class VariablePair
+    {
+    public:
+        VariablePair(std::wstring rawValueIn) :
+            rawValue(std::move(rawValueIn))
+        {
+        }
+        VariablePair() = default;
+        virtual ~VariablePair() = default;
+        std::wstring rawValue;
+        std::wstring resolvedValue;
+    };
+
+    // Function Description:
+    // - Resolves a key's value from within envMap. Uses recursion to recursively resolve
+    //   referenced keys. envMap is updated as values are resolved so we don't double up on
+    //   work
+    // Arguments:
+    // - key: the key that may be in envMap or the environment (or unknown)
+    // - envMap: the map with all the keys, raw and resolved values
+    // - seenKeys: the list of keys that we have already seen when attempting to resolve the
+    //             current chain of keys. seenKeys needs to be passed in as a parameter so
+    //             that each layer in the recursion only has keys seen leading up to it
+    // Return Value:
+    // - The resolved value for key
+    std::wstring ResolveEnvironmentVariableValue(const std::wstring& key, std::map<std::wstring, VariablePair>& envMap, std::list<std::wstring> seenKeys = {})
+    {
+        // First check if the key is in our environment variable map
+        auto it = envMap.find(key);
+        if (it != envMap.end())
+        {
+            if (!it->second.resolvedValue.empty())
+            {
+                // key has already been resolved, so just return it
+                return it->second.resolvedValue;
+            }

```

Comment:
This won't notice if the environment variable has been previously resolved with an empty value. But okay, it's only a shortcut.

Reply:
I could make this more explicit and store something so that we know it has been resolved if you think it is worth the complexity?

Reply:
I think the complexity is not worth it until this becomes so slow that users notice.